### ISSUE TITLE
webp support times out for some images

### ISSUE THREAD

**Issue by rdm** [imgproxy/imgproxy#166]
I have found an image where the webp support in imageproxy takes way too long to handle.

https://img.sunset02.com/sites/default/files/image/2016/07/main/classic-grilled-artichokes-sun-0416.jpg

Could we get a configuration option to set a threshold on the time allowed for compression, like maybe IMGPROXY_ENFORCE_WEBP=15seconds (where, in this example, if it can't convert to webp in 15 seconds it would serve the non-webp variant  of the image)?

**Comment by rdm** [comment]
(Note that the 2.3 release announced in https://github.com/imgproxy/imgproxy/issues/123 should greatly reduce the significance of this issue, so I think that that release is more important.)

**Comment by DarthSim** [comment]
Hi,

As far as I understand, you're trying to convert the image to WebP without any resizing. Saving such a big image is a big job, even for libvips. JPEG is the fastest format here, especially when `libjpeg-turbo` is used, for sure, but all other formats will cause timeout with this image size as well as WebP.

Fallback on JPEG could partially solve the issue, but there are some problems:

1. I can't predict the time of saving since it depends on a lot of things, either I can't stop the process of saving.
2. If not stop the process of saving, imgproxy will continue to do redundant work which will roughly slow down everything else.
3. JPEG needs some work to be done. For example, the image should be flattened with a non-transparent background.

In other words, there are too many complications to make this feature easily done. If you're going to save such big images, you probably should consider increasing the timeout window.

May I know the case where you need to convert such a big picture without resizing? Maybe I would offer some solution for it?

**Comment by rdm** [comment]
That sounds right.

I'm using a two stage process with image proxy, where I first extract the base crop (not resized) from the image, and then resize it. And, I have an environmental variable set which enables webp conversion for browsers which support webp.

And, yes - I've worked around this issue in this setup by specifying that the first stage (cropping) should always generate a jpg format image (this overrides the browser's preferences, but it's ok because the browser isn't seeing that image).

The reason I'm doing this is that I'm generating a srcset for an image crop. But, I couldn't figure out how to generate single-pass imgproxy urls so that my srcset images would all represent the same crop. (Thinking about this, I might be able to adjust the crop boundaries to match what integer arithmetic would give me for resize-first-then-crop for the smallest image and then use corresponding numbers for the larger crops... but I haven't gone there, yet.)

Let's say I have a 5000x5000 base image, and I'm cropping to a 4000x3000 region starting at 123 pixels in from the left side and 456 pixels down from the top of that base image. And, then generating a srcset with these sizes:

400x300
800x600
1200x900

All of these images should represent the same pixels from the original image. But I could not figure out how to accurately specify these images, so I went with a slow but precise two-stage process.

Anyways, as a result, I'm stumbling over performance issues, and this was one of them.

**Comment by DarthSim** [comment]
`darthsim/imgproxy:latest` includes `crop` option that replaces `resizing_type:crop`. Also I added offsets for gravity. Now you can do the following for your example case:

```
/signature/crop:4000:3000:nowe:123:456/resize:fit:400:300/plain/http://example.com/image.jpg
```

**Comment by rdm** [comment]
Oh, very nice!

Thank you.

**Comment by rdm** [comment]
One more question:

Let's say that, instead, my base crop was a simple center crop. Would this be explicitly supported by the new crop: operation? Or do I have to specify null offsets?

**Comment by DarthSim** [comment]
Offsets are optional. If you're going to keep them zeroes, you can just omit them: `crop:ce`

**Comment by DarthSim** [comment]
Closing the issue since the `crop` option is released in v2.3.0. Feel free to open the new one if there is something wrong with `crop`.
